### PR TITLE
feat(client): added provider compatibility check for Driver Adapters

### LINKED ISSUE #21864: [Driver Adapters]: missing provider compatibility validation
Consider [this](https://www.notion.so/prismaio/Driver-Adapters-usage-varying-engineType-Prisma-Client-entrypoints-and-constructor-arguments-c71f234805774898a26c6b2eb307f334) (internal) Notion doc.

Considering this `schema.prisma`:

```prisma
generator client {
  provider = "prisma-client-js"
  engineType = "library"
  previewFeatures = ["driverAdapters"]
}

datasource db {
  provider = "postgres"
  url      = env("DATABASE_URL")
}
```

The schema clearly uses the `postgres` provider.

Now consider a TypeScript snippet in which we load `planetscale`'s Driver Adapter:

```typescript
import { connect } from '@planetscale/database'
import { PrismaPlanetScale } from '@prisma/adapter-planetscale'
import { PrismaClient } from '@prisma/client'

const connectionString = `${process.env.DATABASE_URL}`

const connection = connect({ url: connectionString })
const adapter = new PrismaPlanetScale(connection)
const prisma = new PrismaClient({ adapter })

// use Prisma as usual on any model, e.g., `user`
await prisma.user.findMany()
```

The schema clearly uses the `postgres` provider, yet we are allowed to load the `planetscale`'s Driver Adapter, which should required the `mysql` provider. One can construct a particularly dangerous example in which, as long as the remote `mysql`-based database has the same table definitions as the ones inferred in TypeScript from the Prisma schema, the user could accidentally write to the wrong database, and Prisma Client would give absolutely no indication of this mishap.

So, we're missing a runtime validation error, which could be phrased as:

```
The Driver Adapter `PlanetScale` is not compatible with the `postgres` provider specified in the schema. You should either load a Driver Adapter based on `postgres`, or update the `datasource` block in the schema to use the `mysql` provider instead.
``` 

In particular, one can check that `adapter.flavour === 'mysql'` (see [code](https://github.com/prisma/prisma/blob/8fdb2bcb5b2944d9f4b6dd1a9efa20c88f0b5ab4/packages/adapter-planetscale/src/planetscale.ts#L31)).

**Edge case**: Both `postgresql` and `postgres` are valid `provider`s in the `datasource` block, but only `postgres` is a valid value for the `flavour` property of any `DriverAdapter`' class.

---

Prisma: `5.5.2`

### PR DESCRIPTION
**Pull request by jkomyno**
This PR closes https://github.com/prisma/prisma/issues/21864.

When the Driver Adapter passed to the Prisma Client has an underlying `provider` incompatible with the provider declared in `schema.prisma`, we now throw a runtime validation error. Here's an example when passing `@prisma/adapter-libsql` to a Prisma Client generated from a schema with `provider = "postgres"` / `provider = "postgresql"`:

> "The Driver Adapter \`@prisma/adapter-libsql\`, based on \`sqlite\`, is not compatible with the provider \`postgres\` specified in the Prisma schema."

---

Note: In Prisma 6, we should probably standardise the `provider` values in the `Queryable` interface for Driver Adapters, especially to avoid having two strings for Postgres databases (`postgres`, `postgresql`).
- `getConfig(..).datasources[0].provider` can be `postgresql`, `postgres`, `mysql`, or other known providers
- `getConfig(..).datasources[0].activeProvider`, stored in `config.activeProvider`, can be `postgresql`, `mysql`, or other known providers
- `adapter.provider` can be `postgres`, `mysql`, or `sqlite`, and changing this requires changes to Rust as well,
    see https://github.com/prisma/prisma-engines/blob/d116c37d7d27aee74fdd840fc85ab2b45407e5ce/query-engine/driver-adapters/src/types.rs#L22-L23.

### CODE REVIEW

**Review by Jolg42 — COMMENTED** [review]
@jkomyno It looks like some tests are failing

Example https://github.com/prisma/prisma/actions/runs/9446199134/job/26015382025?pr=24481#step:7:1475
```
FAIL driver-adapters/enforce-preview-feature/tests.ts (8.726 s)
  ● driver-adapters.enforce-preview-feature (provider=postgresql, previewFeatures=[metrics, driverAdapters]) › enforce driverAdapters preview feature

    expect(received).not.toThrow()

    Error name:    "PrismaClientInitializationError"
    Error message: "The Driver Adapter `postgres`, based on `postgres`, is not compatible with the provider `postgresql` specified in the Prisma schema."
```

Another example https://github.com/prisma/prisma/actions/runs/9446199134/job/26015409819?pr=24481#step:5:400
```
  ● typescript › globals.d.ts

    assert.fail(received, expected)

    Message:
      Type '"postgres"' is not assignable to type '"mysql" | "sqlite" | "postgresql"'. Did you mean '"postgresql"'? At: /home/runner/work/prisma/prisma/packages/client/tests/functional/driver-adapters/_utils/provider.ts:12:7

      70 |     for (const checkPath of keys(suiteChecks)) {
      71 |       if (checkPath.includes(path.dirname(suiteFilePath))) {
    > 72 |         assert.fail(suiteChecks[checkPath])
         |                ^
      73 |       }
      74 |     }
      75 |   })

      at typescript/tests.ts:72:16
```

File: packages/client/tests/functional/driver-adapters/validate-active-provider/tests.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+import type { PrismaClient } from './node_modules/@prisma/client'
+
+declare let datasourceInfo: DatasourceInfo
+declare let newPrismaClient: NewPrismaClient<typeof PrismaClient>
+
+// https://github.com/prisma/prisma/issues/21864
+testMatrix.setupTestSuite(
+  ({ driverAdapter, provider }) => {
+    console.log('driverAdapter', driverAdapter)
+    console.log('provider', provider)
```
Comment (Jolg42) [inline packages/client/tests/functional/driver-adapters/validate-active-provider/tests.ts:23]:
```suggestion
```

**Comment by Jolg42** [comment]
@jkomyno looks like good progress! I see an e2e test failing now
https://github.com/prisma/prisma/actions/runs/9450786716/job/26030262470?pr=24481#step:5:4783
```
     '    Snapshot name: `driver adapters cannot be used via @prisma/client/edge 1`\n' +
    '\n' +
    '    - Snapshot  - 2\n' +
    '    + Received  + 1\n' +
    '\n' +
    '    - Prisma Client was configured to use the `adapter` option but it was imported via its `/edge` endpoint.\n' +
    '    - Please either remove the `/edge` endpoint or remove the `adapter` from the Prisma Client constructor.\n' +
    '    + The Driver Adapter `undefined`, based on `undefined`, is not compatible with the provider `postgres` specified in the Prisma schema.\n' +
    '\n' +
```

**Comment by janpio** [comment]
> Note: In Prisma 6, we should probably standardise the `provider` values in the `Queryable` interface for Driver Adapters, especially to avoid having two strings for Postgres databases (`postgres`, `postgresql`).

This is the odd one out:

> * `adapter.provider` can be `postgres`, `mysql`, or `sqlite`, and changing this requires changes to Rust as well,
> see [prisma/prisma-engines@d116c37/query-engine/driver-adapters/src/types.rs#L22-L23](https://github.com/prisma/prisma-engines/blob/d116c37d7d27aee74fdd840fc85ab2b45407e5ce/query-engine/driver-adapters/src/types.rs#L22-L23).

The canonical should match our public default for PostgreSQL, which is `postgresql`. 

(We support `postgres://` in the connection string only as that is what some database providers include in the connection strings they provide to users - and asking our users to fix that would be unnecessary complexity.)

**Comment by jkomyno** [comment]
Hopefully `e2e` tests are fixed now.

**Comment by jkomyno** [comment]
Spotted the problems:
- https://github.com/prisma/prisma/blob/582580ec9a199c525d01a5d9fa76b155ad29f0da/packages/client/tests/e2e/prisma-client-imports/src/dep.ts
- https://github.com/prisma/prisma/blob/582580ec9a199c525d01a5d9fa76b155ad29f0da/packages/client/tests/e2e/prisma-client-imports/src/no-dep.ts

These files are using a `@prisma/client` with `provider = "sqlite"`, yet they're importing any provider and instantiating them within the Prisma Client. I have no real clue about why they're needed here, so we're discussing it via DMs with @Jolg42

File: packages/client/tests/functional/_utils/setupFilesAfterEnv.ts [hunk 2]
```diff
@@ -1,5 +1,7 @@
 // add all jest-extended matchers
 // see https://jest-extended.jestcommunity.dev/docs/matchers/
+import 'undici'
+
```
Comment (Jolg42) [inline packages/client/tests/functional/_utils/setupFilesAfterEnv.ts:4]: Remove? Or add a comment

**Review by Jolg42 — DISMISSED** [review]
Nice!
And we're done with that one 😄

**Comment by jkomyno** [comment]
Looks like e2e tests are failing for the same reason as https://github.com/prisma/prisma/pull/24513#issuecomment-2164919474.

File: packages/client/src/runtime/getPrismaClient.ts [hunk 3]
```diff
@@ -364,6 +362,31 @@ export function getPrismaClient(config: GetPrismaClientConfig) {
...
+      if (optionsArg?.adapter) {
+        adapter = bindAdapter(optionsArg.adapter)
+
+        // Note:
+        // - `getConfig(..).datasources[0].provider` can be `postgresql`, `postgres`, `mysql`, or other known providers
+        // - `getConfig(..).datasources[0].activeProvider`, stored in `config.activeProvider`, can be `postgresql`, `mysql`, or other known providers
+        // - `adapter.provider` can be `postgres`, `mysql`, or `sqlite`, and changing this requires changes to Rust as well,
+        //    see https://github.com/prisma/prisma-engines/blob/d116c37d7d27aee74fdd840fc85ab2b45407e5ce/query-engine/driver-adapters/src/types.rs#L22-L23.
+        //
+        // TODO: Normalize these provider names once and for all in Prisma 6.
```
[the remark this answers is not in the document]
Comment (jkomyno) [reply packages/client/src/runtime/getPrismaClient.ts:379]: https://github.com/prisma/team-orm/issues/1207 + https://www.notion.so/prismaio/Prisma-6-5bcf2bfbb323492d8ab808650719f0ef?pvs=4#55577030daa34ef4867d0a724f08f105

**Review by Jolg42 — APPROVED** [review]
again

**Pull request merged**